web3.js/src/connection.ts

```diff
@@ -2865,39 +2909,72 @@ export class Connection {
       }
     });
 
-    let timeoutMs = this._confirmTransactionInitialTimeout || 60 * 1000;
-    switch (subscriptionCommitment) {
-      case 'processed':
-      case 'recent':
-      case 'single':
-      case 'confirmed':
-      case 'singleGossip': {
-        timeoutMs = this._confirmTransactionInitialTimeout || 30 * 1000;
-        break;
+    const checkBlockHeight = async () => {
+      const blockHeight = await this.getBlockHeight(commitment);
+      return blockHeight;
+    };
+
+    const expireTx = new Promise((resolve, reject) => {
+      if (typeof strategy === 'string') {
+        console.log('timeout');
+        let timeoutMs = this._confirmTransactionInitialTimeout || 60 * 1000;
+        switch (subscriptionCommitment) {
+          case 'processed':
+          case 'recent':
+          case 'single':
+          case 'confirmed':
+          case 'singleGossip': {
+            timeoutMs = this._confirmTransactionInitialTimeout || 30 * 1000;
+            break;
+          }
+          // exhaust enums to ensure full coverage
+          case 'finalized':
+          case 'max':
+          case 'root':
+        }
+        if (done) clearTimeout();
+        setTimeout(() => resolve(TransactionStatus.EXPIRED), timeoutMs);
+      } else if (Object.prototype.hasOwnProperty.call(strategy, 'signature')) {
+        let config = strategy as BlockheightBasedTransactionConfimationStrategy;
+        (async () => {
+          try {
+            let currentBlockHeight = await checkBlockHeight();
+            console.log(currentBlockHeight);
+            console.log(config.lastValidBlockHeight);
+            if (done) return;
+            while (currentBlockHeight <= config.lastValidBlockHeight) {
+              await sleep(1000);
+              if (done) return;
+              currentBlockHeight = await checkBlockHeight();
+              if (done) return;
+            }
+            resolve(TransactionStatus.EXPIRED);
+          } catch (error) {
+            reject(error);
```

Comment:
This is actually something new in your implementation. Previously, the only thing that `confirmTransaction` could really do is to throw an error after the full timeout elapsed. Now, it can fatal _really_ early if the `getBlockHeight` method fatals for any reason:

* connection issues, maybe, or
* the RPC returns a failure response for `getBlockHeight`

While these imply that the block height is not available to the client at the moment, I'm not sure if it implies that the client will _never_ receive the signature confirmation through the websocket. Failure to get the block height might just be transient, and might resolve the very next time you try. Maybe.

So yeah, what do we do? We could:

1. Wait a second and retry. This isn't perfect, because it implies that `expireTx` will last forever if the `checkBlockHeight` method _never_ succeeds.
2. Throw an error if we haven't been able to determine what the block height is after X seconds of trying.
3. Throw an error if we haven't been able to determine what the block height is after X retries.

I'm inclined to suggest that you try #1. It doesn't seem right for a (transient) failure to obtain the blockheight to nerf `confirmTransaction` completely. Let's retry until we get an answer.

cc/ @mvines for thoughts.

Reply:
Hmm, interesting. I'll get to work on it!

Reply:
I haven't had a chance to get into this, but I will.